File: shop/src/main/java/basic/shop/member/command/domain/Member.java
package basic.shop.member.command.domain;

import lombok.Getter;

import javax.persistence.Embedded;
import javax.persistence.EmbeddedId;
import javax.persistence.Entity;
import java.util.Objects;

@Entity
@Getter
public class Member {
    @EmbeddedId
    private MemberId memberId;
    @Embedded
    private Password password;


    public void changePassword(String currentPassword, String newPassword) {
        if(!password.match(currentPassword)) {
            throw new IllegalArgumentException("Not Match Password");
        }
        this.password = new Password(newPassword);
    }
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Member member = (Member) o;
        return Objects.equals(memberId, member.memberId) && Objects.equals(password, member.password);
    }

    @Override
    public int hashCode() {
        return Objects.hash(memberId, password);
    }
}

File: spring-lock/pessimistic-lock/src/main/kotlin/basic/pessimisticlock/member/domain/entity/Member.kt
package basic.pessimisticlock.member.domain.entity

import basic.pessimisticlock.common.model.AbstractTimeEntity
import basic.pessimisticlock.member.domain.value.Age
import basic.pessimisticlock.member.domain.value.MemberId
import javax.persistence.Embedded
import javax.persistence.EmbeddedId
import javax.persistence.Entity
import javax.persistence.Table

@Entity
@Table(name = "member")
class Member (
    @EmbeddedId
    var id: MemberId?,

    var userName: String,

    @Embedded
    var age: Age
): AbstractTimeEntity() {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Member

        if (id != other.id) return false
        if (userName != other.userName) return false
        if (age != other.age) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + userName.hashCode()
        result = 31 * result + age.hashCode()
        return result
    }

    companion object {
        fun generate(id: MemberId, userName: String, age: Age): Member {
            verifyAge(age)
            return Member(id, userName, age)
        }

        private fun verifyAge(age: Age) {
            if(age.value <= 0)
                throw IllegalArgumentException("Not Valid Age")
        }
    }
}